src/rgw/rgw_zone_types.h

```diff
@@ -542,31 +542,77 @@ struct RGWZoneGroupPlacementTierS3 {
 };
 WRITE_CLASS_ENCODER(RGWZoneGroupPlacementTierS3)
 
+enum GlacierRestoreTierType {
+  Standard = 0,
+  Expedited = 1,
+};
+
+struct RGWZoneGroupTierS3Glacier {
+#define DEFAULT_GLACIER_RESTORE_DAYS 1
+  uint64_t glacier_restore_days = DEFAULT_GLACIER_RESTORE_DAYS;
+  GlacierRestoreTierType glacier_restore_tier_type{Standard};
+
+  int update_params(const JSONFormattable& config);
+  int clear_params(const JSONFormattable& config);
+
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(glacier_restore_days, bl);
+    encode(glacier_restore_tier_type, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator& bl) {
+    DECODE_START(1, bl);
+    decode(glacier_restore_days, bl);
+    uint32_t it;
+    decode(it, bl);
+    glacier_restore_tier_type = (GlacierRestoreTierType)it;
+    DECODE_FINISH(bl);
+  }
+  void dump(Formatter *f) const;
+  void decode_json(JSONObj *obj);
+  static void generate_test_instances(std::list<RGWZoneGroupTierS3Glacier*>& o) {
+    o.push_back(new RGWZoneGroupTierS3Glacier);
+    o.back()->glacier_restore_days = 2;
+    o.back()->glacier_restore_tier_type = GlacierRestoreTierType::Expedited;
+  }
+};
+WRITE_CLASS_ENCODER(RGWZoneGroupTierS3Glacier)
+
 struct RGWZoneGroupPlacementTier {
 #define DEFAULT_READ_THROUGH_RESTORE_DAYS 1
 
   std::string tier_type;
   std::string storage_class;
   bool retain_head_object = false;
-  bool allow_read_through = false;
-  uint64_t read_through_restore_days = 1;
 
   struct _tier {
     RGWZoneGroupPlacementTierS3 s3;
   } t;
 
+  bool allow_read_through = false;
+  uint64_t read_through_restore_days = 1;
+  std::string restore_storage_class = RGW_STORAGE_CLASS_STANDARD;
+
+  RGWZoneGroupTierS3Glacier s3_glacier;
+
   int update_params(const JSONFormattable& config);
   int clear_params(const JSONFormattable& config);
 
   void encode(bufferlist& bl) const {
-    ENCODE_START(2, 1, bl);
+    ENCODE_START(4, 1, bl);
     encode(tier_type, bl);
     encode(storage_class, bl);
     encode(retain_head_object, bl);
+    if (is_tier_type_s3()) {
```

Comment:
The tiertype should be encoded in same location right, otherwise encoding/decoding will have issues for upgrades??

Reply:
This change is already done as part of correcting ordering of these options - [PR#61569](https://github.com/ceph/ceph/pull/61569/files#diff-08be2660ab13639bf44e739cf4c4e565d199848c2ed3c49627827350a18bd268R567). That commit is also part of this PR (Waiting for it to get merged).

The only change actually related this PR here is to move the check `if (tier_type == "cloud-s3")` to `if (is_tier_type_s3())`